File: backend-java/FullStackJavaProjectBE/comment-service/src/main/java/be/pxl/services/service/CommentService.java
package be.pxl.services.service;

import be.pxl.services.controller.dto.CommentDTO;
import be.pxl.services.controller.request.AddCommentRequest;
import be.pxl.services.controller.request.EditCommentRequest;
import be.pxl.services.domain.Comment;
import be.pxl.services.exceptions.CommentException;
import be.pxl.services.exceptions.UnautherizedException;
import be.pxl.services.repository.CommentRepository;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class CommentService implements ICommentService{

    private final CommentRepository commentRepository;
    private static final Logger logger = LoggerFactory.getLogger(CommentService.class);


    @Override
    public void addComment(AddCommentRequest request, String userRole, String username, String userId) {

        if (!userRole.equals("user") && !userRole.equals("author") && !userRole.equals("editor")) {
            throw new CommentException("User is not authorized to add a comment");
        }




        Comment comment = Comment.builder()
                .message(request.getMessage())
                .postId(request.getPostId())
                .userId(Long.parseLong(userId))
                .usernameMadeBy(username)
                .dateCreated(LocalDate.now())
                .build();





        commentRepository.save(comment);
        logger.info("Comment has been added");
    }

    @Override
    public List<CommentDTO> getComments(long postId) {
        List<Comment> comments = commentRepository.findAll().stream().filter(comment -> comment.getPostId() == postId).toList();
        return comments.stream()
                .map(comment -> new CommentDTO(comment.getId(),comment.getMessage(), comment.getUsernameMadeBy(), comment.getDateCreated()))
                .toList();
    }

    @Override
    public CommentDTO editComment(long commentId, EditCommentRequest request, String userRole, String username, String userId) {
        if (!userRole.equals("user") && !userRole.equals("author")) {
            throw new UnautherizedException("User is not authorized to edit a comment");
        }

        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new CommentException("Comment not found"));

        if (!(comment.getUserId() == (Long.parseLong(userId)))) {
            throw new UnautherizedException("User is not authorized to edit this comment");
        }
        comment.setMessage(request.getMessage());

        commentRepository.save(comment);
        logger.info("Comment has been edited");


        return new CommentDTO(comment.getId(),comment.getMessage(), comment.getUsernameMadeBy(), comment.getDateCreated());
    }

    @Override
    public void deleteComment(long commentId, String userRole, String username, String userId) {
        if (!userRole.equals("user") && !userRole.equals("author")) {
            throw new UnautherizedException("User is not authorized to delete a comment");
        }

        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new CommentException("Comment not found"));

        if (!(comment.getUserId() == (Long.parseLong(userId)))) {
            throw new UnautherizedException("User is not authorized to delete this comment");
        }

        commentRepository.delete(comment);
        logger.info("Comment has been deleted");
    }
}

File: frontend-web/src/app/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { AddCommentRequest, EditCommentRequest } from './commentrequests';
import { CommentDTO } from './commentdtos';

@Injectable({
  providedIn: 'root',
})
export class CommentService {
  private apiUrl = `${environment.commentApiUrl}`;

  constructor(private http: HttpClient) {}

  addComment(request: AddCommentRequest, userRole: string, username: string, userId: string): Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
      'Role': userRole,
      'User': username,
      'Userid': userId,
    });
    return this.http.post<void>(this.apiUrl, request, { headers });
  }

  getComments(postId: number): Observable<CommentDTO[]> {
    return this.http.get<CommentDTO[]>(`${this.apiUrl}/${postId}`);
  }

  editComment(commentId: number, request: EditCommentRequest, userRole: string, username: string, userId: string): Observable<void> {
    const headers = new HttpHeaders({
      'Content-Type': 'application/json',
      'Role': userRole,
      'User': username,
      'Userid': userId,
    });
    return this.http.post<void>(`${this.apiUrl}/comment/${commentId}`, request, { headers });
  }

  deleteComment(commentId: number, userRole: string, username: string, userId: string): Observable<void> {
    const headers = new HttpHeaders({
      'Role': userRole,
      'User': username,
      'Userid': userId,
    });
    return this.http.delete<void>(`${this.apiUrl}/comment/${commentId}`, { headers });
  }
}
